File: code_cpp/Hashing/L4_Unordered_set_in_C++STL.cpp
#include<iostream>
#include<unordered_set>
using namespace std;

int main(){
    unordered_set<int> s;
    s.insert(10);
    s.insert(51);
    s.insert(5);
    s.insert(20);
    s.insert(30);
    s.insert(24);
    // unoredered_set doesnot maintain order of element
    for(int x:s)
        cout<<x<<" ";
    // An other way to iterate
    cout<<"\nother way to iterate\n";
    for(auto it=s.begin(); it!=s.end(); it++)
    {
        cout<<(*it)<<" ";
    }
    cout<<endl;
    // find() function
    if(s.find(51)==s.end())
        cout<<"Not found"<<endl;
    else
        cout<<"found "<<*(s.find(51))<<endl;
    // count() function--> count function can be used of find function
    // Alternative of found function
    // count() function, when element present return 1 else 0 
    if(s.count(20))
        cout<<"Found"<<endl;
    else
        cout<<"Not Found"<<endl;
    // Erase function
    cout<<"Size before erase "<<s.size()<<endl;
    s.erase(5);
    cout<<"Size after erase "<<s.size()<<endl;
    // Erase using other way
    auto it=s.find(24);
    s.erase(it);
    cout<<"Size after again erase "<<s.size()<<endl;
    // we can also use erase function to erase set of element
    s.erase(s.begin(),s.end());
    cout<<"Size after fully erasing "<<s.size()<<endl;
    //clear the array
    s.clear();// erase everything
    cout<<"After clearing size "<<s.size()<<endl;
    
    return 0;
}

File: code_java/self pased DSA/Hashing/L4_Unordered_set_in_C++STL.java
/*
Problem: Demonstrate basic operations of unordered_set (C++)/HashSet (Java) such as insertion, iteration, search, erase, and size.
Approach:
- Use Java's HashSet to perform similar operations as C++ unordered_set.
*/
import java.util.*;

public class L4_Unordered_set_in_Java {
    public static void main(String[] args) {
        Set<Integer> s = new HashSet<>();
        s.add(10);
        s.add(51);
        s.add(5);
        s.add(20);
        s.add(30);
        s.add(24);
        // HashSet does not maintain order
        for (int x : s)
            System.out.print(x + " ");
        System.out.println("\nOther way to iterate");
        for (Iterator<Integer> it = s.iterator(); it.hasNext(); ) {
            System.out.print(it.next() + " ");
        }
        System.out.println();
        // find() equivalent: contains
        if (!s.contains(51))
            System.out.println("Not found");
        else
            System.out.println("found 51");
        // count() equivalent: contains
        if (s.contains(20))
            System.out.println("Found");
        else
            System.out.println("Not Found");
        System.out.println("Size before erase " + s.size());
        s.remove(5);
        System.out.println("Size after erase " + s.size());
        s.remove(24);
        System.out.println("Size after again erase " + s.size());
        s.clear();
        System.out.println("After clearing size " + s.size());
    }
} 